rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+# 004 - Better Policy UX RFC
+
+## Current Status
+
+### Proposed
+
+2022-09-06
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Full Name <email@example.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Joseph <joseph@iasql.com>
+
+## Summary
+
+Customers (even the very technically experienced ones) often have to look up how to construct an AWS IAM policy and then spend time validating the policy before inserting into IaSQL.
+This becomes more and more unmanageable as the policy grows.
+This is not a good experience, and it can be improved upon. We believe we can collect some inputs from IaSQL customers and offer then some form of automation when creating policies, so they can do it faster, more confidently and more conveniently.
+
+## Proposal
+
+We should add new `Policy` and `Statement` entities (perhaps in the `aws_iam` module). These entities would serve as 
+a way to construct IAM policies in a more user-friendly way. For example, the customer intends to create an S3 bucket with
+read-only permissions from all Principals within the same AWS account. Ideally, something like this:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Resource": "aws:arn:123456789012:bucket-resource",
+      "Action": [
+        "s3:GetObjects"
+      ],
+      "Principal": {
+        "AWS": "arn:aws:iam::123456789012:root"
+      }
+    }
+  ]
+}
+```
+
+Instead of the customer having to look up and put together a JSON blob like this and then verifying and then inserting, we can also give them
+the option to do this using functions that would generate this blob.
+
+```sql
+-- User creates policy
+select * from iasql_create_policy(
+    'some_policy_name',
+    '2012-10-17' -- nullable, defaults to 2012-10-17
+    );
+
+-- User creates statement and attaches to the policy
+select * from iasql_create_policy_statement(
+        '<policy_id>',
+        'allow',
+        'aws:arn:123456789012:bucket-resource',
+        's3:GetObjects', -- comma separated list of actions
+        '123456789012' -- comma separated list of principals
```

Comment:
The example above shows the Principals as a key-value structure. This doesn't seem possible?

Reply:
My thinking here was to have the postgres function generate something like this
```
{
  "AWS": [
          "22232323232",
          "56677889765"
        ]
}
```
when multiple strings are provided. The way it is on https://awspolicygen.s3.amazonaws.com/policygen.html

Reply:
So digging through [the documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/access_policies.html) it says

> Principal (Required in only some circumstances) – If you create a resource-based policy, you must indicate the account, user, role, or federated user to which you would like to allow or deny access. If you are creating an IAM permissions policy to attach to a user or role, you cannot include this element. The principal is implied as that user or role.

It's not a property that is always necessary (in fact, most of the examples do not include this).

There are also other properties not being modeled here, particularly `sid` and `condition`, where `condition` [can have some very strange JSON embedded in it](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_aws_deny-requested-region.html) that this proposal can't model.

Reply:
I have addressed these in the second draft.
Condition always takes the form of 
```
{
  "<Condition>": {
     "<Key>": "<Value>"
  }
}
```
I represented this as an array of triples